src/Features/LanguageServer/Protocol/Handler/ServerLifetime/LspServiceLifeCycleManager.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.ServerLifetime;
+
+internal class LspServiceLifeCycleManager : ILifeCycleManager, ILspService
```

Comment:
This isn't in CLaSP, it's a Roslyn internal deal. This is because you suggested that rather than making the members of `ILifeCycleManager` virtual on AbstractLanguageServer this should be accomplished via writing your own ILifeCycleManager. So I wrote this one which takes an AbstractLanguageServer in the constructor to accomplish the Roslyn-specific work in shutdown.